File: packetCapture/src/main/java/com/ninpou/packetcapture/core/util/common/LruCache.java
package com.ninpou.packetcapture.core.util.common;

import java.util.LinkedHashMap;

public class LruCache<K, V> extends LinkedHashMap<K, V> {
    private int maxSize;
    private transient CleanupCallback<V> callback;

    public LruCache(int maxSize, CleanupCallback<V> callback) {
        super(maxSize + 1, 1, true);

        this.maxSize = maxSize;
        this.callback = callback;
    }


    @Override
    protected boolean removeEldestEntry(Entry<K, V> eldest) {
        if (size() > maxSize) {
            callback.cleanUp(eldest.getValue());
            return true;
        }
        return false;
    }

    public interface CleanupCallback<V> {
        void cleanUp(V v);
    }
}

File: packetCapture/src/main/java/com/ninpou/packetcapture/core/util/common/LruCache.kt
package com.ninpou.packetcapture.core.util.common

import java.util.*

class LruCache<K, V>(private val maxSize: Int, @field:Transient private val callback: CleanupCallback<V>) : LinkedHashMap<K, V>(maxSize + 1, 1f, true) {
    override fun removeEldestEntry(eldest: Map.Entry<K, V>): Boolean {
        if (size > maxSize) {
            callback.cleanUp(eldest.value)
            return true
        }
        return false
    }

    interface CleanupCallback<V> {
        fun cleanUp(v: V)
    }
}